src/main/java/org/opensearch/ad/ml/ModelManager.java

```diff
@@ -822,15 +649,23 @@ public void maintenance(ActionListener<Void> listener) {
                 models.remove(modelId);
             }
             if (modelState.getLastCheckpointTime().plus(checkpointInterval).isBefore(now)) {
-                checkpointDao.putModelCheckpoint(modelId, toCheckpoint.apply(modelState.getModel()), ActionListener.wrap(r -> {
+                ActionListener<Void> checkpointListener = ActionListener.wrap(r -> {
                     modelState.setLastCheckpointTime(now);
-                    maintenanceForIterator(models, toCheckpoint, iter, listener);
+                    maintenanceForIterator(models, iter, listener);
                 }, e -> {
                     logger.warn("Failed to finish maintenance for model id " + modelId, e);
-                    maintenanceForIterator(models, toCheckpoint, iter, listener);
-                }));
+                    maintenanceForIterator(models, iter, listener);
+                });
+                T model = modelState.getModel();
+                if (model instanceof RandomCutForest) {
+                    checkpointDao.putRCFCheckpoint(modelId, (RandomCutForest) model, checkpointListener);
+                } else if (model instanceof ThresholdingModel) {
+                    checkpointDao.putThresholdCheckpoint(modelId, (ThresholdingModel) model, checkpointListener);
+                } else {
+                    maintenanceForIterator(models, iter, listener);
```

Comment:
This should never be reached correct?

Reply:
yeah, sent IllegalArgumentException instead.